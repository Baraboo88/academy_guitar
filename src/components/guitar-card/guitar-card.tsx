import React, {useState} from 'react';
import {CartItemModel, GuitarModel} from '../../types/guitar-model';
import {
  generateStars,
  getCyrillicRating,
  getPriceWithSpaces,
  getAdapterImage,
  handlerCartItemIncrease
} from '../../utils/utils';
import {Link} from 'react-router-dom';
import AddToCartModal from '../add-to-cart-modal/add-to-cart-modal';
import {StateModel} from '../../types/redux-models';
import {getCartItems} from '../../store/cart/cart-selector';
import {ThunkDispatch} from 'redux-thunk';
import {RootState} from '../../index';
import {AxiosStatic} from 'axios';
import {Action} from 'redux';
import {CartActionCreator} from '../../store/cart/cart-actions';
import {connect} from 'react-redux';
import AddToCartModalSuccess from '../add-to-cart-modal-success/add-to-cart-modal-success';

export enum StarSize{
    Main, CardDetails, Comments
}


export const renderStars = (rating: number, size: StarSize) => generateStars().map((el, index) => {

  switch(size) {
    case StarSize.CardDetails:
      return (
        <svg key={`${rating} - ${el}`} width="14" height="14" aria-hidden="true">
          <use xlinkHref={rating >= el ? '#icon-full-star' : '#icon-star'}></use>
        </svg>);
    case StarSize.Main:
      return     (
        <svg key={`${rating} - ${el}`} width="12" height="11" aria-hidden="true">
          <use xlinkHref={rating >= el ? '#icon-full-star' : '#icon-star'}></use>
        </svg>);
    case StarSize.Comments:
      return (
        <svg key={`${rating} - ${el}`} width="16" height="16" aria-hidden="true">
          <use xlinkHref={rating >= el ? '#icon-full-star' : '#icon-star'}></use>
        </svg>);
    default:
      return <div></div>;
  }


});

interface GuitarCardProps{
  card: GuitarModel;
  cartItems: CartItemModel [];
  setCartItems: (cartItems: CartItemModel []) => void;
}

function GuitarCard(props :GuitarCardProps) {
  const {card, cartItems, setCartItems} = props;
  const [isAddToCardPopUpOpened, setIsAddToCardPopUpOpened] = useState(false);
  const [isAddToCardPopUpOpenedSuccess, setIsAddToCardPopUpOpenedSuccess] = useState(false);
  const handlerIsAddToCardClose = () => {
    setIsAddToCardPopUpOpened(false);
  };
  const handlerIsAddToCardCloseSuccess = () => {
    setIsAddToCardPopUpOpenedSuccess(false);
  };

  const handlerCartItemAdd = () => {
    setIsAddToCardPopUpOpened(false);
    setIsAddToCardPopUpOpenedSuccess(true);
    setCartItems(handlerCartItemIncrease(cartItems, card));
  };

  return (
    <>
      <div className="product-card">
        <img src={`/img/content/catalog-product-${getAdapterImage(card.previewImg)}.jpg`}
          srcSet={`/img/content/catalog-product-${getAdapterImage(card.previewImg)}@2x.jpg 2x`}
          width="75"
          height="190" alt={`${card.name}`}
        />
        <div className="product-card__info">
          <div className="rate product-card__rate">
            {renderStars(card.rating, StarSize.Main)}
            <p className="visually-hidden">Рейтинг: {getCyrillicRating(card.rating)}</p>
            <p className="rate__count">
              <span
                className="visually-hidden"
              >Всего оценок:
              </span> {card.comments ? card.comments.length : 0}
            </p>
          </div>
          <p className="product-card__title">{card.name}</p>
          <p className="product-card__price">
            <span className="visually-hidden">Цена:</span>
            {getPriceWithSpaces(card.price)} ₽
          </p>
        </div>
        <div className="product-card__buttons">

          <Link to={`/product/${card.id}`} className="button button--mini">Подробнее
          </Link>

          {cartItems.find((cartItem) => cartItem.guitar.id === card.id) === undefined ?
            <button
              onClick={() => {
                setIsAddToCardPopUpOpened(true);
              }}
              className="button button--red button--mini button--add-to-cart"
            >Купить
            </button> :
            <Link to={'/cart'} className="button button--red-border button--mini button--in-cart">В Корзине</Link>}

        </div>

      </div>
      {isAddToCardPopUpOpened && <AddToCartModal onAddToCard={handlerCartItemAdd} onModalClose={handlerIsAddToCardClose} guitar={card}/>}
      {isAddToCardPopUpOpenedSuccess && <AddToCartModalSuccess onModalClose={handlerIsAddToCardCloseSuccess}/>}
    </>
  );
}

const mapStateToProps = (state: StateModel) => ({
  cartItems: getCartItems(state),
});


const mapDispatchToProps = (dispatch: ThunkDispatch<RootState, AxiosStatic, Action>) => ({
  setCartItems(cartItems: CartItemModel []) {
    dispatch(CartActionCreator.setCartItems(cartItems));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(GuitarCard);

export {GuitarCard};
