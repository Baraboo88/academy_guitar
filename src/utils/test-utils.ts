import {AddCommentModel, GuitarCommentModel, GuitarModel, GuitarType} from '../types/guitar-model';
import {createApi} from '../api';
import {applyMiddleware, createStore} from 'redux';
import thunk from 'redux-thunk';
import {ReactWrapper} from 'enzyme';


export const MOCK_ERROR= 'Somme error';

export const mockGuitars: GuitarModel[] = [
  {
    id: 1,
    name: 'Честер Bass',
    vendorCode: 'SO757575',
    type: 'electric' as GuitarType,
    description: 'Замечательный малобюджетный вариант, созданный для новичков, которые отдают предпочтение мелодичным стилям. Прекрасно звучат блюз и баллады, исполненные на этой гитаре. Акустические свойства весьма высоки, в отличие от ее стоимости.',
    previewImg: 'img/guitar-1.jpg',
    stringCount: 7,
    rating: 4,
    price: 17500,
  },
  {
    id: 2,
    name: 'CURT Z300',
    vendorCode: 'TK129049',
    type: 'electric' as GuitarType,
    description: 'Эргономичность гитары и качество сборки являются, пожалуй, её главными преимуществами. Идеальное расположение в руках музыканта дополняется прочностью конструкции из клёна.',
    previewImg: 'img/guitar-8.jpg',
    stringCount: 7,
    rating: 3.5,
    price: 29500,
  },
  {
    id: 3,
    name: 'Roman LX',
    vendorCode: 'RO111111',
    type: 'ukulele' as GuitarType,
    description: 'Укулеле класса премиум от компании CURT, собравшая в себе все самые необходимые качесва: лёгкость корпуса, прочность струн и компактный размер.',
    previewImg: 'img/guitar-6.jpg',
    stringCount: 4,
    rating: 4,
    price: 6800,
  },
  {
    id: 4,
    name: 'CURT T300',
    vendorCode: 'TK436457',
    type: 'electric' as GuitarType,
    description: 'CURT T300 - это шестиструнная электрогитара популярной линейки FPT. Модель с классическим стилем головы грифа и деки.',
    previewImg: 'img/guitar-3.jpg',
    stringCount: 6,
    rating: 5,
    price: 30000,
  },
  {
    id: 5,
    name: 'Dania Super',
    vendorCode: 'DI192138',
    type: 'acoustic' as GuitarType,
    description: 'Гитары производителя Dania пользуются популярностью у музыкантов разного уровня. В модели Super идеально сочетаются демократичная цена и качество. Корпус, выполненный из тополя позволяет корпусу долгое время сохранять свой первоначальный вид.',
    previewImg: 'img/guitar-4.jpg',
    stringCount: 7,
    rating: 4.5,
    price: 3500,
  },
  {
    id: 6,
    name: 'Честер WX',
    vendorCode: 'SO934345',
    type: 'electric' as GuitarType,
    description: 'Электрогитара с олд-скульным грифом и глянцевым корпусом для настоящих рок-музыкантов. Насыщенный звук идеально подойдет для исполнения рок-композиций.',
    previewImg: 'img/guitar-2.jpg',
    stringCount: 6,
    rating: 3.5,
    price: 15300,
  },
  {
    id: 7,
    name: 'Dania VX',
    vendorCode: 'DI082347',
    type: 'ukulele' as GuitarType,
    description: 'Укулеле - прекрасный компактный аналог классической акустической гитары. Прекрасное качество продуктов компании Dania сочетается с демократичными ценами продуктов линейки GSR.',
    previewImg: 'img/guitar-5.jpg',
    stringCount: 4,
    rating: 2.5,
    price: 2200,
  },
];
export const testInitialState = {
  currentGuitar: {
    currentGuitar: mockGuitars,
    errorMsg: '',
  },
  guitars: {
    guitars: mockGuitars,
    errorMsg: '',
  },
};

export const mockCommentToAdd: AddCommentModel = {
  advantage: 'Mock Advantage', comment: 'Mock Comment', disadvantage: 'Mock Disadvantage', guitarId: 3, rating: 3, userName: 'Some name',
};


export const mockComment: GuitarCommentModel = {id: '31b10247-adff-44a7-937e-c71af64f425b',
  userName: 'Никита',
  advantage: 'Цена.',
  disadvantage: 'Быстро приходит в негодность покрытие краской.',
  comment: 'Хорошая гитара для начинающих, сделана качественно. Лучше за эти деньги не найти.',
  rating: 5,
  createAt: new Date('2022-03-10T10:56:01.590Z'),
  guitarId: 1};

export const getTestStore = () => {

  const reducer = (state = testInitialState) => state;
  const api = createApi();

  return createStore(reducer, applyMiddleware(thunk.withExtraArgument(api)));
};

export const findByTestAtr = (component:ReactWrapper, attr: string) => component.find(`[data-test='${attr}']`);
