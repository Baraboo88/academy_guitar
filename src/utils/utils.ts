const STARTS_AMOUNT = 5;

export const imageAdapter = (img: string) => img.slice(img.length - 5, img.length - 4);


export const generateStarsArray = () => {
  const stars = [];
  for (let i = 1; i <= STARTS_AMOUNT; i++) {
    stars.push(i);
  }
  return stars;
};

export const getCyrillicRating = (rating: number) => {
  if(rating> 0 && rating < 3){
    return 'Плохо';
  } else if (rating >= 3 && rating < 4){
    return 'Удовлетворительно';
  } else if(rating >= 4 && rating < 5){
    return 'Хорошо';
  } else {
    return 'Отлично';
  }
};
