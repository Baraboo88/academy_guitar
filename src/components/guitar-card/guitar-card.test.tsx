import * as renderer from 'react-test-renderer';
import * as React from 'react';


import {GuitarCard} from './guitar-card';
import { mockCartItems, mockGuitars} from '../../utils/test-utils';
import { BrowserRouter } from 'react-router-dom';


it('GuitarCard successfully rendered', () => {
  const tree = renderer.create(  <BrowserRouter><GuitarCard card={mockGuitars[0]} setCartItems={jest.fn()} cartItems={mockCartItems}/></BrowserRouter>);
  expect(tree).toMatchSnapshot();
});
