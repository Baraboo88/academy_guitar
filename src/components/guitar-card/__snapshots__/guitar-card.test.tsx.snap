// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuitarCard successfully rendered 1`] = `
<div
  className="product-card"
>
  <img
    alt="Честер Bass"
    height="190"
    src="/img/content/catalog-product-1.jpg"
    srcSet="/img/content/catalog-product-1@2x.jpg 2x"
    width="75"
  />
  <div
    className="product-card__info"
  >
    <div
      className="rate product-card__rate"
    >
      <svg
        aria-hidden="true"
        height="11"
        width="12"
      >
        <use
          xlinkHref="#icon-full-star"
        />
      </svg>
      <svg
        aria-hidden="true"
        height="11"
        width="12"
      >
        <use
          xlinkHref="#icon-full-star"
        />
      </svg>
      <svg
        aria-hidden="true"
        height="11"
        width="12"
      >
        <use
          xlinkHref="#icon-full-star"
        />
      </svg>
      <svg
        aria-hidden="true"
        height="11"
        width="12"
      >
        <use
          xlinkHref="#icon-full-star"
        />
      </svg>
      <svg
        aria-hidden="true"
        height="11"
        width="12"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
      <p
        className="visually-hidden"
      >
        Рейтинг: 
        Хорошо
      </p>
      <p
        className="rate__count"
      >
        <span
          className="visually-hidden"
        >
          Всего оценок:
        </span>
         
        0
      </p>
    </div>
    <p
      className="product-card__title"
    >
      Честер Bass
    </p>
    <p
      className="product-card__price"
    >
      <span
        className="visually-hidden"
      >
        Цена:
      </span>
      17 500
       ₽
    </p>
  </div>
  <div
    className="product-card__buttons"
  >
    <a
      className="button button--mini"
      href="/product/1"
      onClick={[Function]}
    >
      Подробнее
    </a>
    <a
      className="button button--red-border button--mini button--in-cart"
      href="/cart"
      onClick={[Function]}
    >
      В Корзине
    </a>
  </div>
</div>
`;
